### PR TITLE
Improved support `apply from` in gradle files

### PR DESCRIPTION
**Pull request by zbynek**
For now work in progress, should fix https://github.com/dependabot/dependabot-core/issues/2138

I still need to add some tests etc, for now I'm just wondering if there is any interest in getting this merged as a stop-gap solution before the Gradle Tooling is integrated as proposed in https://github.com/dependabot/dependabot-core/issues/1164

### CODE REVIEW

**Comment by zbynek** [comment]
@greysteil do you think it's worth it to finish and merge this? I understand #1164 is a better approach, but I don't have the knowledge or time needed to implement that and there seems to be no progress on that issue.

**Comment by greysteil** [comment]
Hey @zbynek, I'm not the right person to ask! @asciimike / @hmarr are likely to have a better view.

**Comment by asciimike** [comment]
@JLLeitschuh might be a good candidate to review this, if available; otherwise, someone on our eng team will take a look at this when free.

**Comment by JLLeitschuh** [comment]
A set of unit/integration tests would help me better understand what you're trying to do here

File: gradle/lib/dependabot/gradle/file_parser.rb [hunk 1]
```diff
@@ -298,13 +298,17 @@ def buildfiles
         end
       end
 
+      def self.find_includes(buildfile, dependency_files)
+        buildfile.content.
+          scan(/apply from(\s+=|:)\s+['"]([^'"]+)['"]/).flatten.
```
Comment (JLLeitschuh) [inline gradle/lib/dependabot/gradle/file_parser.rb:303]: This won't work for the Kotlin DSL as the syntax is different. Also, I don't think this covers the full Groovy syntax that's supported for this declaration.

**Comment by JLLeitschuh** [comment]
You may want to check for recursion and make sure that your logic doesn't get stuck checking two files that are potentially self-referential.

**Comment by zbynek** [comment]
@JLLeitschuh thanks for checking this! I've added some tests. 

There aim of this PR is not to implement full support for `apply`, only to add support for some usecases.

Improvements in this PR:
* resolve property defined in included file and used in (sub)project build file
* resolve repositories defined in a file included from the main project file
* basic Kotlin syntax `apply(from = "dependencies.gradle.kts")`

Existing limitations not addressed by this PR:
* no support for more advanced syntaxes (e.g. filename defined by property, Koltin `apply` with `mapOf` etc.)
* no recursive resolution of includes
* no support for `apply to:` (`to` is ignored)
* all included files that are actually fetched must contain `dependencies` in the filename

**Comment by JLLeitschuh** [comment]
> all included files that are actually fetched must contain dependencies in the filename

That will require some kind of documentation somewhere too

**Comment by JLLeitschuh** [comment]
I don't think you should drop this PR, but I do want to point you towards the more idiomatic way of doing this with Gradle.

We're encouraging our users to move away from using `apply from:` to using either plugins, or using platforms.

https://docs.gradle.org/current/userguide/platforms.html

**Comment by JLLeitschuh** [comment]
The test look sane, so a 👍 for me from here.

**Comment by zbynek** [comment]
@JLLeitschuh thanks. For the docs I created https://github.com/github/docs/pull/10041/files .

**Comment by zbynek** [comment]
> The test look sane, so a 👍 for me from here.

@asciimike given the approval from @JLLeitschuh , what are the next steps to get this merged?

**Review by jurre — APPROVED** [review]
Thanks for the contribution, the code looks good to me with the updated docs 👍

**Pull request merged**